src/uct/tcp/tcp_ep.c

```diff
@@ -225,39 +241,120 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
     ucs_trace_data("tcp_ep %p: recvd %zu bytes", ep, recv_length);
 
     /* Parse received active messages */
-    while ((remainder = ep->length - ep->offset) >= sizeof(*hdr)) {
-        hdr = ep->buf + ep->offset;
+    while (ep->offset < ep->length) {
+        remainder = ep->length - ep->offset;
+        if (remainder < sizeof(*hdr)) {
+	    return recv_length > 0;
+        }
+
+        hdr = ep->buf + ep->offset; 
         ucs_assert(hdr->length <= (iface->config.buf_size - sizeof(uct_tcp_am_hdr_t)));
 
         if (remainder < sizeof(*hdr) + hdr->length) {
-            break;
+            return recv_length > 0;
         }
 
         /* Full message was received */
         ep->offset += sizeof(*hdr) + hdr->length;
 
-        if (hdr->am_id >= UCT_AM_ID_MAX) {
-            ucs_error("invalid am id: %d", hdr->am_id);
-            continue;
+        uct_tcp_ep_complete_recv_am(iface, ep, hdr);
+    }
+
+    ep->length = ep->offset = 0;
+    ucs_mpool_put_inline(ep->buf);
+
+    return recv_length > 0;
+}
+
+static inline
+unsigned uct_tcp_ep_do_partial_rx(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    ucs_status_t status;
+    uct_tcp_am_hdr_t *hdr;
+    size_t cur_recvd_length = ep->length - ep->offset;
+    size_t recv_length;
+
+    if (cur_recvd_length < sizeof(*hdr)) {
+        recv_length = sizeof(*hdr) - cur_recvd_length;
+
+        status = uct_tcp_recv(ep->fd, ep->buf + ep->length, &recv_length);
+        if (status != UCS_OK) {
+            if (status == UCS_ERR_CANCELED) {
+                ucs_debug("tcp_ep %p: remote disconnected", ep);
```

Comment:
can do common cleanup function (and check whether need to return buf to mpool there)